### PR TITLE
Uri/fix vgp wrapper

### LINKED ISSUE #406: TF compilation causes problems for VGP models
**Describe the bug**
TF compilation causes problems for models other than GPR when used with EI

**To reproduce**
To reproduce, we start with the EI notebook: https://secondmind-labs.github.io/trieste/notebooks/expected_improvement.html
In this notebook, replace GRP with VGP, by replacing this code:
```
def build_model(data):
    variance = tf.math.reduce_variance(data.observations)
    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=[0.2, 0.2])
    prior_scale = tf.cast(1.0, dtype=tf.float64)
    kernel.variance.prior = tfp.distributions.LogNormal(tf.cast(-2.0, dtype=tf.float64), prior_scale)
    kernel.lengthscales.prior = tfp.distributions.LogNormal(tf.math.log(kernel.lengthscales), prior_scale)
    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
    gpflow.set_trainable(gpr.likelihood, False)

    return GPflowModelConfig(**{
        "model": gpr,
        "model_args": {
            "num_kernel_samples": 100,
        },
        "optimizer": gpflow.optimizers.Scipy(),
        "optimizer_args": {
            "minimize_args": {"options": dict(maxiter=100)},
        },
    })
```

with this (the smallest possible change to replace GPR with VGP):
```
def build_model(data):
    variance = tf.math.reduce_variance(data.observations)
    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=[0.2, 0.2])
    prior_scale = tf.cast(1.0, dtype=tf.float64)
    kernel.variance.prior = tfp.distributions.LogNormal(tf.cast(-2.0, dtype=tf.float64), prior_scale)
    kernel.lengthscales.prior = tfp.distributions.LogNormal(tf.math.log(kernel.lengthscales), prior_scale)
    likelihood = gpflow.likelihoods.Gaussian()
    vgp = gpflow.models.VGP(data.astuple(), kernel, likelihood)

    return GPflowModelConfig(**{
        "model": vgp,
        "optimizer": gpflow.optimizers.Scipy(),
        "optimizer_args": {
            "minimize_args": {"options": dict(maxiter=100)},
        },
    })
```

Now run it and observe the following error:
> tensorflow.python.framework.errors_impl.FailedPreconditionError:  Could not find variable _AnonymousVar18. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status=Not found: Resource localhost/_AnonymousVar18/N10tensorflow3VarE does not exist.

(not pasting full stack trace in interest of readability)

**Expected behaviour**
I would expect the notebook to execute without errors.

**System information**
 - Trieste version: top of develop
 - TensorFlow version: 2.6.0
 - GPflow version: 2.2.1

**"Fixes"**
Not suggesting any of these as fixes, but as evidence that this is a TF compilation problem:

1. If we don't compile EI function call by removing decorator `@tf.function` from it (see [line link](https://github.com/secondmind-labs/trieste/blob/develop/trieste/acquisition/function.py#L249)) the error goes away.
2. If we don't update EI function, and create a new one every time, like we do anyway for most of acquisition functions (to do that replace `function.update` [here](https://github.com/secondmind-labs/trieste/blob/develop/trieste/acquisition/function.py#L220) with `return expected_improvement(model, eta)` as [here](https://github.com/secondmind-labs/trieste/blob/develop/trieste/acquisition/function.py#L201)), the error goes away.

### PR DESCRIPTION
**Pull request by uri-granta**
Fix for https://github.com/secondmind-labs/trieste/issues/406. The issue was that the VGP variational parameters `VGP.m_qu` and `VGP.m_sqrt`, which are stored in `gpflow.Parameters`, need to be `.assigned` to like Variables rather than overriden.

Unfortunately to be able to assign we need to give them variable shapes, just like we do for the data. However `gpflow.Parameter` doesn't support this, so I've hacked a modified version that does (which in turn requires a modified version of `tfp.python.util.TransformedVariable`).

In the case where `use_natgrads` is False there is still an issue if people try to use the Adam optimizer, as the optimizer maintains state (e.g. moment vectors) which can't handle the changing parameter shapes. Suggest we handle this in a separate PR. For now I've removed tests of this combination (which isn't the recommended default).

### CODE REVIEW

File: tests/integration/test_bayesian_optimization.py [hunk 1]
```diff
@@ -176,19 +181,44 @@ def test_optimizer_finds_minima_of_simple_quadratic(
 ) -> None:
     # for speed reasons we sometimes test with a simple quadratic defined on the same search space
     # branin; currently assume that every rule should be able to solve this in 5 steps
-    _test_optimizer_finds_minimum(False, min(num_steps, 5), acquisition_rule)
+    _test_optimizer_finds_minimum(min(num_steps, 5), acquisition_rule)
+
+
+@random_seed
+@pytest.mark.parametrize("use_natgrads", [False, True])
+def test_optimizer_with_vgp_model(use_natgrads: bool) -> None:
```
Comment (hstojic) [inline tests/integration/test_bayesian_optimization.py:197]: should we rename the main test above then to mention gpr?

File: tests/integration/test_bayesian_optimization.py [hunk 2]
```diff
@@ -198,9 +228,30 @@ def build_model(data: Dataset) -> GaussianProcessRegression:
...
+        elif model_type == "SVGP":
+            Z = search_space.sample_sobol(20)  # Initialize diverse inducing locations
+            svgp = gpflow.models.SVGP(
+                kernel,
+                gpflow.likelihoods.Gaussian(variance=1e-5),
+                Z,
+                num_data=len(data.observations),
+            )
+            gpflow.utilities.set_trainable(svgp.likelihood, False)
+            return SparseVariational(svgp, BatchOptimizer(tf.optimizers.Adam(0.1)))
```
Comment (hstojic) [inline tests/integration/test_bayesian_optimization.py:260]: seems like a high learning rate for adam, but if the tests pass I guess its fine

File: tests/unit/models/gpflow/test_models.py [hunk 3]
```diff
@@ -55,6 +55,7 @@
     SparseVariational,
     VariationalGaussianProcess,
 )
+from trieste.models.gpflow.models import NumDataPropertyMixin
```
Comment (hstojic) [inline tests/unit/models/gpflow/test_models.py:59]: there's been quite a few updates in this file in latest PR

File: trieste/models/gpflow/models.py [hunk 4]
```diff
@@ -469,9 +479,70 @@ def num_data(self, value: TensorType) -> None:
         self._num_data.assign(value)
 
 
-class SVGPWrapper(SVGP, NumDataPropertyMixin):
-    """A wrapper around GPFlow's SVGP class that stores num_data in a tf.Variable and exposes
-    it as a property."""
+class Parameter(gpflow.Parameter):
```
Comment (hstojic) [inline trieste/models/gpflow/models.py:489]: can we move this class to util.py in this folder? this file is getting rather big...

File: trieste/models/gpflow/models.py [hunk 5]
```diff
@@ -496,7 +567,12 @@ def __init__(self, model: SVGP, optimizer: Optimizer | None = None):
         # GPflow stores num_data as a number. However, since we want to be able to update it
         # without having to retrace the acquisition functions, put it in a Variable instead.
         # So that the elbo method doesn't fail we also need to turn it into a property.
-        if not isinstance(self._model, SVGPWrapper):
+        if not isinstance(self._model, NumDataPropertyMixin):
+
+            class SVGPWrapper(type(self._model), NumDataPropertyMixin):  # type: ignore
```
Comment (hstojic) [inline trieste/models/gpflow/models.py:581]: better solution I think

File: trieste/models/gpflow/models.py [hunk 6]
```diff
@@ -601,27 +672,67 @@ def __init__(
 
         self._use_natgrads = use_natgrads
         self._natgrad_gamma = natgrad_gamma
+        self._ensure_variable_model_data()
+
+    def _ensure_variable_model_data(self) -> None:
```
Comment (hstojic) [inline trieste/models/gpflow/models.py:692]: this whole check could potentially live in util.py as well, takes a single argument (model) and optionally returns three objects with which we update some parts of it

**Review by hstojic — COMMENTED** [review]
with the caveat that I couldn't assess properly the content of Parameter object and related changes in that _ensure method, it looks ok to me, 2 main comments:
1. I think most of this could go to utils.py, these classes and the file are getting big, and these things are not essential for understanding the VGP class so I don't think we lose much by moving them out of the view
2. not only related to the current PR, but as a more general solution for our sequential decision making use case would be to raise an issue/PR in GPflow, it might be possible to allow for dynamic shapes on their side, without hurting the regular supervised learning use case

File: tests/unit/models/gpflow/test_models.py [hunk 7]
```diff
@@ -547,6 +549,11 @@ def test_gaussian_process_regression_optimize(
     else:
         args = {}
     loss = internal_model.training_loss(**args)
+
+    # VariationalGaussianProcess doesn't currently support Adam() unless use_natgrads is True
+    if isinstance(internal_model, VGP) and isinstance(optimizer, tf.optimizers.Optimizer):
+        return
```
Comment (jesnie) [inline tests/unit/models/gpflow/test_models.py]: Maybe use `pytest.skip`, so we leave a log of this not being executed?

**Review by jesnie — APPROVED** [review]
LGTM

File: trieste/models/gpflow/models.py (hunk 4 above)
Comment (apaleyes) [reply trieste/models/gpflow/models.py:489]: Late to the party, but wanted second this. Perhaps just separate parameter.py in the same folder would be appropriate for this class?

**Pull request merged**